lib/promscrape/discovery/consul/watch.go

```diff
@@ -78,11 +81,59 @@ func (cw *consulWatcher) mustStop() {
 	// TODO: add ability to cancel blocking requests.
 }
 
+func (cw *consulWatcher) updateServices(serviceNames []string) {
+	// Start watchers for new services.
+	for _, serviceName := range serviceNames {
+		cw.servicesLock.Lock()
+		if _, ok := cw.services[serviceName]; ok {
+			// The watcher for serviceName already exists.
+			continue
+		}
+		sw := &serviceWatcher{
+			serviceName: serviceName,
+			stopCh:      make(chan struct{}),
+		}
+		cw.services[serviceName] = sw
+		cw.servicesLock.Unlock()
+
+		cw.wg.Add(1)
+		serviceWatchersCreated.Inc()
+
+		initCh := make(chan struct{})
+		go func() {
+			serviceWatchersCount.Inc()
+			sw.watchForServiceNodesUpdates(cw, initCh)
+			serviceWatchersCount.Dec()
+			cw.wg.Done()
+		}()
+		// wait for initialization to complete
+		<-initCh
```

Comment:
This may take long time for big number of `serviceNames`. It would be better moving the wait for initialization for all the concurrently started serviceWatchers outside the loop.

Reply:
Thanks! Moved initialization block out of loop in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/2202/commits/563a0e8de6785cfa7057a4769ccc095de62afe26